[AIRFLOW-3741] Add extra config to Oracle hook

zhongjiajie opened the pull request:
Add extra config to Oracle hook
including encoding, mode, threaded etc

Add Oracle hook unittest

Add docs in insert_rows and
bulk_insert_rows

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3741
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Add extra config to Oracle hook

### Tests

- [x] My PR adds the following unit tests : 

test_oracle_hook.py

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

ashb commented on airflow/hooks/oracle_hook.py:
```diff
@@ -50,20 +50,61 @@ def get_conn(self):
-            conn = cx_Oracle.connect(conn.login, conn.password, conn.host)
+            conn_config['dsn'] = conn.host
 
+        if conn.extra_dejson.get('encoding', None):
```
```suggestion
        if "encoding" in conn.extra_dejson:
```

And similar changes for all of the other `if get(x. None)` conditions.

zhongjiajie replied:
fix

ashb commented on airflow/hooks/oracle_hook.py:
```diff
@@ -50,20 +50,61 @@ def get_conn(self):
+                conn_config['nencoding'] = conn.extra_dejson.get('encoding')
+        if conn.extra_dejson.get('nencoding', None):
+            conn_config['nencoding'] = conn.extra_dejson.get('nencoding')
+        mode = conn.extra_dejson.get('mode', None)
```
```suggestion
        mode = conn.extra_dejson.get('mode', '').lower()
```

And then we don't need `lower()` on all of the branches, or the extra `if mode:` on line 79

zhongjiajie replied:
That great

ashb commented on airflow/hooks/oracle_hook.py:
```diff
@@ -50,20 +50,61 @@ def get_conn(self):
+            elif mode.lower() == 'sysrac':
+                conn_config['mode'] = cx_Oracle.SYSRAC
+        if conn.extra_dejson.get('threaded', None):
+            conn_config['threaded'] = True
```
```suggestion
            conn_config['threaded'] = conn.extra_dejson['threaded']
```

And similar for events.

zhongjiajie replied:
@ashb I do that because parameters only accept `True` or `False`, I wonder user will pass other value

ashb replied:
I wouldn't worry about that. More confusing would be a user specifying `"threaded": false` and have it turn on!

Just document it (or point at the oracle docs) as boolean and keep it simple on our side

zhongjiajie replied:
Ok, I get it

ashb commented on airflow/hooks/oracle_hook.py:
```diff
@@ -128,13 +181,33 @@ def bulk_insert_rows(self, table, rows, target_fields=None, commit_every=5000):
         cursor = conn.cursor()
-        values = ', '.join(':%s' % i for i in range(1, len(target_fields) + 1))
-        prepared_stm = 'insert into {tablename} ({columns}) values ({values})'.format(
+        if target_fields:
```
This change looks unrelated to the config change. what's this for?

zhongjiajie replied:
@ashb This change ref the sub commit message
> Fix Oracle hook bulk_insert_rows bug
when target_fields use default None
value

Old code define `target_fields=None` but when target_fields is None both 
https://github.com/apache/airflow/blob/2c7bb17435926628914571f14dad3c45bc53753b/airflow/hooks/oracle_hook.py#L136
and 
https://github.com/apache/airflow/blob/2c7bb17435926628914571f14dad3c45bc53753b/airflow/hooks/oracle_hook.py#L139
will raise Error

So this is to fix that

ashb replied:
Your PR is "Add extra config to Oracle hook" - that sounds like a different issue. Please could you open that as a separate PR?

zhongjiajie replied:
Ok, I will submit new PR to do that.

ashb commented on tests/hooks/test_oracle_hook.py:
```diff
@@ -0,0 +1,248 @@
+import unittest
+from datetime import datetime
+
+import cx_Oracle
```
Could you guard this in an try/except and decorate the whole class with a `@unittest.skipIf`?

See tests_mongo_hook for an example (but only one decorator on the whole class, not one per fn please)

ashb commented on airflow/hooks/oracle_hook.py:
```diff
@@ -50,20 +50,61 @@ def get_conn(self):
         conn = self.get_connection(self.oracle_conn_id)
+        conn_config = {
+            'user': conn.login,
+            'password': conn.password or ''
```
Is empty string different to None? I'd probably say the `or ''` isn't needed, but not certain with Oracle.

zhongjiajie replied:
You're right, my fault.

kaxil commented on airflow/hooks/oracle_hook.py:
```diff
@@ -48,22 +48,65 @@ def get_conn(self):
         """
         conn = self.get_connection(self.oracle_conn_id)
+        conn_config = {
+            'user': conn.login,
```
Maybe add a condition block to check if `login` & `password` fields are not empty and raise an error that they are required if it is empty.

zhongjiajie replied:
@kaxil When I write code. I think empty usr or pwd would raise by cx_Oracle.connect in L109. What do you think. Should I write raise error block?

ashb replied:
This isn't a regression or a change in behaviour, so I'll merge this.

zhongjiajie committed:
[AIRFLOW-3741] Add extra config to Oracle hook

Add extra config to Oracle hook
including encoding, mode, threaded etc

Add Oracle hook unittest

Add docs in insert_rows bulk_insert_rows
and howto manage-connections.rst

ashb committed:
Update oracle_hook.py

ashb committed:
Merge branch 'master' into AIRFLOW_3741

ashb committed:
Update manage-connections.rst
